### ISSUE TITLE
fix Last Respects BP when Terastallized

### ISSUE THREAD

**Issue by doshidak** [doshidak/showdex#135]
> **Reported by:** [**Sorrica**](https://smogon.com/forums/threads/showdex-an-auto-updating-damage-calculator-built-into-showdown.3707265/post-9520801)

[*Last Respects*](https://smogon.com/dex/sv/moves/last-respects) worked as intended with the [manual override on Calcdex's end](https://github.com/doshidak/showdex/issues/103), but now that the [Tera STAB BP boost mechanic](https://bulbapedia.bulbagarden.net/wiki/Terastal_phenomenon#Effects) (which brings any STAB move with < 60 BP to 60 when Terastallized) has been [natively implemented into `@smogon/calc`](https://github.com/smogon/damage-calc/commit/118b01df3b77a9f9f8f773afc8088cb4a7ee4b2f), the manually boosted BP is overwritten to 60, which deflates the resulting damages.

<table>
  <tbody>
    <tr>
      <td align="center"><strong>Before Tera</strong></td>
      <td align="center"><strong>After Tera</strong></td>
    </tr>
    <tr>
      <td><img alt="calcdex-last-respects-pre-tera" src="https://user-images.githubusercontent.com/18254287/232183871-68fa7161-180b-45e9-a7b1-1afa13f0fab1.png" /></td>
      <td><img alt="calcdex-last-respects-post-tera" src="https://user-images.githubusercontent.com/18254287/232183886-b8760b82-f9a4-4a30-a683-16a8a62c4ac7.png" /></td>
    </tr>
  </tbody>
</table>

Also rather curiously, the manually boosted BP (from the move's tooltip) goes from 250 to 300 after Terastallizing. (Though, due to the internal mechanic, damages are calculated at 60 BP, as you can see from the matchup tooltip.)

**Comment by doshidak** [comment]
Figured out why the BP of *Last Respects* is 300 in the Calcdex when Terastallized: the manual implementation of the Tera STAB BP mechanic (`shouldBoostTeraStab()`) still exists in the utility that actually performs the manual boost (`calcMoveBasePower()`).

However, `shouldBoostTeraStab()` runs first, outputting `60`, then calculates the boosted BP of *Last Respects* with the `faintCounter` of the passed-in `pokemon` (e.g., `4` in the example screenshots). So with the math of `basePower * (1 + faintCounter)`, which in this case, becomes `60 * (1 + 4) = 60 * 5 = 300`.

300 BP is actually incorrect since the [Smogon Dex entry for *Last Respects*](https://smogon.com/dex/sv/moves/last-respects) says:

> Power is equal to 50+(X*50), where X is the total number of times any Pokemon has fainted on the user's side, and X cannot be greater than 100.

As for how the interaction is supposed to go, no idea on that front (need to do some research into how Showdown handles it).

Do we:

* Boost the BP first to 60, then add an additional +50 BP for each fainted ally (e.g., resulting in 260 BP), or
* Don't boost the BP at all once the resulting BP after applying the effects from each fainted ally is > 60 (e.g., resulting in 250 BP)?

I'm thinking it's the former, since the formula could look like `basePower + (50 * faintCounter)` (clamping `faintCounter` to a max value of `100`). But I'll still check how Showdown's server calculates the move's final BP, just to make sure.

**Comment by doshidak** [comment]
It appears from [Showdown's server source code](https://github.com/smogon/pokemon-showdown) that it's the latter: don't boost the BP as the resulting BP is > 60.

From [`sim/battle-actions.ts`](https://github.com/smogon/pokemon-showdown/blob/ca8c863097279132819fdd7db0104eabf6bcf9ff/sim/battle-actions.ts#L1583-L1586), we can see that the `basePower` is initially set from `move.basePower`, but calls the move's `basePowerCallback()` function if it exists:

```ts
let basePower: number | false | null = move.basePower;
if (move.basePowerCallback) {
  basePower = move.basePowerCallback.call(this.battle, source, target, move);
}
```

*Last Respects* has a `basePowerCallback()` defined in [`data/moves.ts`](https://github.com/smogon/pokemon-showdown/blob/ca8c863097279132819fdd7db0104eabf6bcf9ff/data/moves.ts#L10211-L10226):

```ts
export const Moves: {[moveid: string]: MoveData} = {
  // ...
  lastrespects: {
    num: 854,
    accuracy: 100,
    basePower: 50,
    basePowerCallback(pokemon, target, move) {
      return 50 + 50 * pokemon.side.totalFainted;
    },
    category: "Physical",
    name: "Last Respects",
    pp: 10,
    priority: 0,
    flags: {protect: 1, mirror: 1},
    secondary: null,
    target: "normal",
    type: "Ghost",
  },
  // ...
};
```

A couple lines below the `basePowerCallback()` call in `moves.ts`, we can find the [implementation of the Tera STAB BP boost mechanic](https://github.com/smogon/pokemon-showdown/blob/ca8c863097279132819fdd7db0104eabf6bcf9ff/sim/battle-actions.ts#L1626-L1632):

```ts
if (
  basePower < 60 && source.getTypes(true).includes(move.type) && source.terastallized && move.priority <= 0 &&
  // Hard move.basePower check for moves like Dragon Energy that have variable BP
  !move.multihit && !((move.basePower === 0 || move.basePower === 150) && move.basePowerCallback)
) {
  basePower = 60;
}
```

From the example above, once `basePowerCallback()` is called, `basePower` is `250`, which fails the first `basePower < 60` conditional in the mechanic implementation.

In other words, the BP is **not** set to `60`, but remains at `250`.

**Comment by doshidak** [comment]
Fixed! (...I hope LOL. Looks right tho!)

<img alt="calcdex-last-respects-post-fix" width="500px" src="https://user-images.githubusercontent.com/18254287/232675180-2daaac0d-5e76-4e2c-b391-e502923341c9.png" />

Verified this on [calc.ps](https://calc.pokemonshowdown.com) after importing the *exact* spreads (from Gen 9 Randoms) & overriding the BP of *Last Respects* to 250:

<table>
  <tbody>
    <tr>
      <td align="center"><strong><em>Basculegion-F</em></strong></td>
      <td align="center"><strong><em>Gyarados</em></strong></td>
    </tr>
    <tr>
      <td>
        <pre>
Basculegion-F (F) @ Choice Scarf
Ability: Adaptability
Tera Type: Ghost
Level: 73
EVs: 84 HP / 84 Atk / 84 Def / 84 SpA / 84 SpD / 84 Spe
Hardy Nature
- Hydro Pump
- Ice Beam
- Last Respects
- Wave Crash
        </pre>
      </td>
      <td>
        <pre>
Gyarados (F) @ Heavy-Duty Boots
Ability: Intimidate
Tera Type: Ground
Level: 79
EVs: 84 HP / 84 Atk / 84 Def / 84 SpA / 84 SpD / 84 Spe
Hardy Nature
- Dragon Dance
- Earthquake
- Stone Edge
- Waterfall
        </pre>
      </td>
    </tr>
  </tbody>
</table>

![smogon-calc-last-respects-verification](https://user-images.githubusercontent.com/18254287/232675401-5e80c12a-74fb-4920-970d-b4ca04f661a0.png)

<img alt="smogon-calc-last-respects-bp-override" width="350px" src="https://user-images.githubusercontent.com/18254287/232675455-ffaa93fa-cbf0-4069-a2ad-f6bac39db2de.png" />

Will close this via reference on the next PR.
